### ISSUE TITLE
Shoot single ray and retrieve the face it hits

### ISSUE THREAD

**Issue by robinenrico** [rnd-team-dev/plotoptix#11]
I would like to shoot rays with a specific origin and angle and retrieve the face it hits from a mesh in the scene. Is this currently possible?

**Comment by robertsulej** [comment]
Shooting single rays is not supported now. If your use case needs this (or maybe a list of rays to shoot together to have a better performance) - it would be straight forward to me to add such camera mode. That may be an interesting option. Reading hit details is already handled and also easy to extend (run e.g. [this example](https://github.com/rnd-team-dev/plotoptix/blob/master/examples/1_basic_plot_making/5_create_basic_mesh.py) and hover the mouse over the object).

**Comment by robinenrico** [comment]
Thanks for your fast reply, Ideally, I would love to shoot a bundle of rays from a specified origin under an angle. For each ray, I would like to know the faceID that is hit by each ray. By this, you would shoot a bundle of x rays toward an object in 3D and you will receive the faces/triangles that are affected.

I guess that for my use case, it might be possible to work with the existing implementations. Do you mean with your last sentence that I will already be able to retrieve the faces from the 3D object on which the light hits? e.g. in your example, you have a pyramid and some faces are light up and others not, is it possible to retrieve the IDs from those faces? Without doing this graphically by hovering over it?

**Comment by robertsulej** [comment]
Now you can use 2D image coordinates (pixel x and y index in the output array) to get:
- id of the vertex nearest to the hit,
- hit 3D position and distance to the camera.

This can be done without GUI, with NpOptiX class, but there is no convenience function.  You need to adopt code from [_gui_get_object_at()](https://github.com/rnd-team-dev/plotoptix/blob/master/plotoptix/tkoptix.py#L259) to check which (if any) object is at given 2D position, [_gui_get_hit_at()](https://github.com/rnd-team-dev/plotoptix/blob/master/plotoptix/tkoptix.py#L250) to have the hit 3D and distance. Tk GUI is decoding it in a few lines [here](https://github.com/rnd-team-dev/plotoptix/blob/master/plotoptix/tkoptix.py#L273). Methods used there self._optix.get_hit_at() and self._optix.get_object_at(), and the dictionary self.geometry_names are available without GUI in NpOptiX.

Use the pinhole camera, e.g. in the mentioned example code:

`optix.setup_camera("cam1", cam_type="Pinhole", eye=[1, -6, 4], target=[0, 0, 0], fov=25)`

I'll add to NpOptiX convenience methods to get this info. Also, vertex id is not yet the face id, so I'll need to add something to access face id from the library under the Python API.

Camera mode with a list of rays to shoot sounds interesting. Need to think a bit how to organize inputs, and will add that as well. Next release should be ~next weekend, with some more additions.

**Comment by aluo-x** [comment]
Ideally it would expose something very similar to TriMesh's [intersects_location api](https://trimsh.org/trimesh.ray.ray_pyembree.html#trimesh.ray.ray_pyembree.RayMeshIntersector.intersects_location). Basically it takes as input a list of ray origins, ray directions; it returns a list of (m, 3) intersection locations, where m is the number of rays that hit a mesh, the indices of the m rays, and the indices of the triangles that it hits.  Or empty arrays if no rays are hit. I believe only the first hit is valid, as the multi-hit option is known to give invalid results (search for bugs in the Trimesh bug tracker). 

My use case is having this ray tracer in the loop of training a neural network. Trimesh's one is CPU accelerated using embree, but unfortunately it does not support batched processing. I plan on investigating plotoptix, and looking at the overhead (context setup) and investigate potential batched processing. 

The downside of embree is that it that it requires a GPU -> CPU transfer, semi-sequential per-mesh ray tracing (already vectorized, so mutliprocessing yields very limited speedups), and another CPU -> GPU transfer.

**Comment by robinenrico** [comment]
I totally agree with @aluo-x . I previously used the TriMesh package for the purpose of finding a ray intersection with triangles. However, I noticed the slow speed as I need to shoot billions of rays in my evolutionary optimization algorithm. I am currently investigating the TriMesh implementation, which is already vectorized. I am currently trying to implement a Pytorch version of the TriMesh intersects_location functions as it seems to be reasonably easy to transform the computations to the GPU with Pytorch due to the vectorized implementation of TriMesh. I currently have the ray intersection with triangles implemented on the GPU with Pytorch. However, I still need to figure out a way to implement the KD tree or R tree on the GPU  as this will dramatically reduce the number of triangles to evaluate. Not sure if this will be possible in full, but I hope to involve the GPU as much as possible in this.

**Comment by aluo-x** [comment]
@robinenrico Very cool! Your task seems to be much more compute intensive than mine. I limit myself to maybe 10K~ rays max. 
I have also considered using [Taichi](https://github.com/taichi-dev/taichi) to perform the ray tracing via CUDA, as it can integrate with PyTorch  (with gradients), but claims higher performance, but haven't really had time to investigate. 

I would be very interested in seeing your batched ray intersection code if the code is public.

**Comment by robertsulej** [comment]
Leaving plotoptix API aside.. OptiX may be what you are looking for. Ray-triangle intersections and BVH are well optimized and run on RT cores if available.

**Comment by robinenrico** [comment]
> Leaving plotoptix API aside.. OptiX may be what you are looking for. Ray-triangle intersections and BVH are well optimized and run on RT cores if available.

Thanks for the idea robert, I looked into this and didn't find a good way to shoot single rays to find the triangle intersection. If this is supported in some way, this will be amazing for most applications. Do you know a way that this can be done using OptiX?

**Comment by aluo-x** [comment]
So one possible approach, without going into the nitty gritty of Optix & CUDA kernels.
mitsuba2 (with Optix/RTX support) provides a python API, which has functions to [compute per ray mesh intersections](https://mitsuba2.readthedocs.io/en/latest/generated/render_api.html#mitsuba.render.Shape.compute_surface_interaction), which returns an [Interaction object](https://mitsuba2.readthedocs.io/en/latest/generated/render_api.html#mitsuba.render.Interaction3f) which stores the intersection point `p` and the [triangle idx](https://mitsuba2.readthedocs.io/en/latest/generated/render_api.html#mitsuba.render.SurfaceInteraction3f.prim_index). Unfortunately the documentation is a bit sparse, so I'm not sure if you can batch submit multiple rays. 

Also @robertsulej apologies for taking the thread off track from plotoptix. I can also discuss with @robinenrico offline if this is too distracting.

**Comment by aluo-x** [comment]
Seems like you can! From [mitsuba2's documentation](https://mitsuba2.readthedocs.io/en/latest/src/python_interface/rendering_scene.html#custom-rendering-pipeline-in-python):
```
rays, weights = sensor.sample_ray_differential(
    time=0,
    sample1=sampler.next_1d(),
    sample2=pos * scale,
    sample3=0
)

# Intersect rays with the scene geometry
surface_interaction = scene.ray_intersect(rays)
After computing the surface intersections for all the rays, we then extract the depth values

# Given intersection, compute the final pixel values as the depth t
# of the sampled surface interaction
result = surface_interaction.t

# Set to zero if no intersection was found
result[~surface_interaction.is_valid()] = 0
```
For GPU usage I would probably use `gpu_rgb` instead of `packet_rgb`. 

No batched processing, and I think no multi-GPU support? I suspect the context setup cost is also pretty high. But you may be able to re-use the GPU context and just modify the meshes. 
If you want to discuss this offline, I can be reached at `afluo [a;t] andrew.cmu.edu`

**Comment by robertsulej** [comment]
@aluo-x no worries :)

@robinenrico 
You need to write a ray generation program. You can control all the ray parameters there, and how many rays to shoot. Then you need a very simple "closest hit" program which writes the face id and hit coordinates to the ray payload. That part is simple, but putting all together into the pipeline is rather tedious. Docs are [here](https://raytracing-docs.nvidia.com/optix7/index.html), and code samples are included in SDK.

**Comment by robinenrico** [comment]
@robertsulej Thanks for your reply, I will take a look into it. Although I'm not really an expert in Cuda so probably this will be a tedious process. 
@aluo-x Mitsuba seems to do the trick indeed. The example looks really clean and understandable. But I will look into that in more detail tomorrow. Regarding the multi-gpu, it does not seem to be supported but I think we can work with this by dividing the workload explicitly over different GPUs and call the full example for each GPU in separation. This is not as clean but probably does the trick for our use cases. In my case, I have my main evolution running on the CPU and if I find a way to explicitly mention the GPU to use for the ray tracing in Mitsuba, than I will be able to setup n programs in parallel where each program has its own rays and GPU.  Maybe, you are able to something similar, as your gradients do not depend on the ray tracing results. But I'm not sure how this will affect your overall performance as your main program uses the GPU already. I will contact you on you mentioned mail to discuss this further.

**Comment by robertsulej** [comment]
I still need to update/debug Linux binaries, but the Windows version is already usable on GitHub. The new release should be on PyPI today or tomorrow. Sorry for delay, commercial work was pressing..

I guess you may already have found a good package for your problem. Anyway you gave me an interesting idea. The implementation fits the plotoptix framework and it was also possible to skip major part of calculations if e.g. only the face id is needed. You can check [this example](https://github.com/rnd-team-dev/plotoptix/blob/master/examples/1_basic_plot_making/12_custom_camera_projection-mesh_variant.ipynb) for the fast face id reading, and [this one](https://github.com/rnd-team-dev/plotoptix/blob/master/examples/1_basic_plot_making/11_output_buffers.ipynb) for accessing other output data.

**Comment by aluo-x** [comment]
Very cool! I just took a look at the face id reading example, and it looks extremely clean. The custom projection API would also be very useful. 

For now, I'm using an embree based solution with object based parallelization. I understand that someone wrote a python hook into optix prime called [pyoptixprime](https://gist.github.com/tim37021/7b95481caf5f10620f8de64695ebe33c), I've reached out to the author but he thinks more time is needed before a public release.

**Comment by robertsulej** [comment]
Great! 👍 

I'm not sure what are the plans for OptiX Prime. It appeared with the OptiX release 6.5, but not with 7.0/7.1. Fortunately 6.5 supports RTX features so for a short term project it looks OK. But for a longer living code it may happen that Prime is discontinued.

**Comment by robertsulej** [comment]
If you have more ideas, feel free to share it :)
For accessing face index it seems usable now - closing the issue.
